### PR TITLE
TaskThrowFix

### PR DESCRIPTION
**Pull request by dshe**
If WriteLogEntry() is called after tests have completed, an InvalidOperationException may be thrown.
To guard against this, exceptions can be swallowed/ignored.

```
The active test run was aborted. Reason: Unhandled Exception: System.AggregateException: An error occurred while writing to logger(s). (There is no currently active test.) ---> System.InvalidOperationException: There is 
no currently active test.
At line:1 char:1
+ dotnet test .\RxSockets.Tests\RxSockets.Tests.csproj
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (The active test...ly active test.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
   at Xun
it.Sdk.TestOutputHelper.GuardInitialized() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestOutputHelper.cs:line 51
   at Xunit.Sdk.TestOutputHelper.QueueTestOutput(String output) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestOu
tputHelper.cs:line 62
   at Divergic.Logging.Xunit.TestOutputLogger.WriteLogEntry[TState](LogLevel logLevel, EventId eventId, TState state, String message, Exception exception, Func`3 formatter) in 
D:\a\1\s\Divergic.Logging.Xunit\TestOutputLogger.cs:line 
91
   at Divergic.Logging.Xunit.FilterLogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter) in D:\a\1\s\Divergic.Logging.Xunit\FilterLogger.cs:line 46
   at Microsoft.Extensions.Logging.Logger.Log[TSt
ate](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.Logging.Logger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception 
exception, Func`3 formatter)
   at Microsoft.Extensions.Logging.Logger`1.Microsoft.Extensions.Logging.ILogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   at Microsoft.Extensions.Logging.LoggerEx
tensions.Log(ILogger logger, LogLevel logLevel, EventId eventId, Exception exception, String message, Object[] args)
   at Microsoft.Extensions.Logging.LoggerExtensions.LogTrace(ILogger logger, String message, Object[] args)
   at RxSockets.RxSocketServe
r.<>c__DisplayClass7_1.<CreateAcceptObservable>b__1() in C:\projects\rxsocket\RxSockets\RxSocketServer.cs:line 57
   at System.Reactive.Concurrency.Scheduler.Invoke(IScheduler scheduler, Action action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurr
ency\Scheduler.Simple.cs:line 253
   at System.Reactive.Concurrency.EventLoopScheduler.Run() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\EventLoopScheduler.cs:line 354
   at System.Threading.Thread.ThreadMain_ThreadStart()
   at System.Thr
eading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw
()
Unhandled Exception: System.AggregateException: An error occurred while writing to logger(s). (There is no currently active test.) ---> System.InvalidOperationException: There is no currently active test.
   at Xunit.Sdk.TestOutputHelper.GuardInitiali
zed() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestOutputHelper.cs:line 51
   at Xunit.Sdk.TestOutputHelper.QueueTestOutput(String output) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestOutputHelper.cs:line 62
   at Divergic
.Logging.Xunit.TestOutputLogger.WriteLogEntry[TState](LogLevel logLevel, EventId eventId, TState state, String message, Exception exception, Func`3 formatter) in D:\a\1\s\Divergic.Logging.Xunit\TestOutputLogger.cs:line 91
   at Divergic.Logging.Xunit.Filt
erLogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter) in D:\a\1\s\Divergic.Logging.Xunit\FilterLogger.cs:line 46
   at Microsoft.Extensions.Logging.Logger.Log[TState](LogLevel logLevel, EventId event
Id, TState state, Exception exception, Func`3 formatter)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.Logging.Logger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   at M
icrosoft.Extensions.Logging.Logger`1.Microsoft.Extensions.Logging.ILogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   at Microsoft.Extensions.Logging.LoggerExtensions.Log(ILogger logger, LogLevel
 logLevel, EventId eventId, Exception exception, String message, Object[] args)
   at Microsoft.Extensions.Logging.LoggerExtensions.LogTrace(ILogger logger, String message, Object[] args)
   at RxSockets.RxSocketServer.<>c__DisplayClass7_1.<CreateAcceptO
bservable>b__1() in C:\projects\rxsocket\RxSockets\RxSocketServer.cs:line 57
   at System.Reactive.Concurrency.Scheduler.Invoke(IScheduler scheduler, Action action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Simple.cs:line 253
  
 at System.Reactive.Concurrency.EventLoopScheduler.Run() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\EventLoopScheduler.cs:line 354
   at System.Threading.Thread.ThreadMain_ThreadStart()
   at System.Threading.ExecutionContext.RunInternal(E
xecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
Unhandled Exception: System.Aggre
gateException: An error occurred while writing to logger(s). (There is no currently active test.) ---> System.InvalidOperationException: There is no currently active test.
   at Xunit.Sdk.TestOutputHelper.GuardInitialized() in C:\Dev\xunit\xunit\src\xunit
.execution\Sdk\Frameworks\TestOutputHelper.cs:line 51
   at Xunit.Sdk.TestOutputHelper.QueueTestOutput(String output) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\TestOutputHelper.cs:line 62
   at Divergic.Logging.Xunit.TestOutputLogger.Write
LogEntry[TState](LogLevel logLevel, EventId eventId, TState state, String message, Exception exception, Func`3 formatter) in D:\a\1\s\Divergic.Logging.Xunit\TestOutputLogger.cs:line 91
   at Divergic.Logging.Xunit.FilterLogger.Log[TState](LogLevel logLeve
l, EventId eventId, TState state, Exception exception, Func`3 formatter) in D:\a\1\s\Divergic.Logging.Xunit\FilterLogger.cs:line 46
   at Microsoft.Extensions.Logging.Logger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception
, Func`3 formatter)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.Logging.Logger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   at Microsoft.Extensions.Logging.Logger`1.
Microsoft.Extensions.Logging.ILogger.Log[TState](LogLevel logLevel, EventId eventId, TState state, Exception exception, Func`3 formatter)
   at Microsoft.Extensions.Logging.LoggerExtensions.Log(ILogger logger, LogLevel logLevel, EventId eventId, Exception
 exception, String message, Object[] args)
   at Microsoft.Extensions.Logging.LoggerExtensions.LogTrace(ILogger logger, String message, Object[] args)
   at RxSockets.RxSocketServer.<>c__DisplayClass7_1.<CreateAcceptObservable>b__1() in C:\projects\rxsoc
ket\RxSockets\RxSocketServer.cs:line 57
   at System.Reactive.Concurrency.Scheduler.Invoke(IScheduler scheduler, Action action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Simple.cs:line 253
   at System.Reactive.Concurrency.Event
LoopScheduler.Run() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\EventLoopScheduler.cs:line 354
   at System.Threading.Thread.ThreadMain_ThreadStart()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, Con
textCallback callback, Object state)
```

### CODE REVIEW

**Comment by roryprimrose** [comment]
I get what you are trying to do here, but it is really the test that is invalid. Any code that executes after the test is finished (logging or otherwise) would be very dangerous. I'm inclined not to take this in unless there is a valid business/logical scenario. The risk is that it could hide a problem with the test that really should be resolved instead. The error being thrown here is correct.

```
The active test run was aborted. Reason: Unhandled Exception: System.AggregateException: An error occurred while writing to logger(s). (There is no currently active test.) ---> System.InvalidOperationException: There is 
no currently active test.
```

**Comment by dshe** [comment]
My program writes to the logger on a background thread. Background threads
are supposed to abort when the main program exits. THis should not allow
any further logging. But it does. So I suspect that the problem is with
Xunit not properly aborting associated background threads before finishing
tests. This is not a new issue:

https://github.com/aspnet/templating/issues/514
https://github.com/xunit/xunit/issues/1540

The exception is thrown on line 51 here:
https://github.com/xunit/xunit/blob/master/src/xunit.execution/Sdk/Frameworks/TestOutputHelper.cs#L51

The following Xunit-logging component seems somehow not to have the problem:
https://github.com/martincostello/xunit-logging/tree/master/src/Logging.XUnit

A workaround would be appreciated.

Thanks for your help!

File: Divergic.Logging.Xunit/LoggingConfig.cs [hunk 1]
```diff
@@ -7,21 +12,63 @@
...
-            Formatter = new DefaultFormatter();
-        }
+        public virtual bool IgnoreTestBoundaryException { get; set; } = false;
 
         /// <summary>
-        ///     Gets or sets a custom formatting for rendering log messages to xUnit test output.
+        ///     Identifies the number of spaces to use for indenting scopes.
         /// </summary>
-        public ILogFormatter Formatter { get; set; }
+        public int PaddingSpaces { get; set; } = 3;
```
Comment (roryprimrose) [inline Divergic.Logging.Xunit/LoggingConfig.cs]: This is a good idea.

File: Divergic.Logging.Xunit/LoggingConfig.cs [hunk 2]
```diff
@@ -7,21 +12,63 @@
...
         /// </summary>
-        public bool IgnoreTestBoundaryException { get; set; }
+        /// <param name="scopeLevel">The number of active logging scopes.</param>
+        /// <param name="name">The logger name.</param>
+        /// <param name="logLevel">The log level.</param>
+        /// <param name="eventId">The event id.</param>
+        /// <param name="message">The log message.</param>
+        /// <param name="exception">The exception to be logged.</param>
+        /// <returns>The formatted log message.</returns>
+        public virtual string Format(
```
Comment (roryprimrose) [inline Divergic.Logging.Xunit/LoggingConfig.cs]: This breaks Single Responsibility Pattern. Formatting is an action that in itself is not configuration. Please revert to how it was using an ILogFormatter property.

You could still combine these concepts with your own custom config class though. Perhaps something like this that also provides a simple singleton access property.

```
public class MyCustomConfig : LoggingConfig, ILogFormatter
{
    private MyCustomConfig()
    {
        Formatter = this;
    }

    public static MyCustomConfig Current { get; } = new MyCustomConfig();
    
    public string Format(
        int scopeLevel,
        string name,
        LogLevel logLevel,
        EventId eventId,
        string message,
        Exception exception)
    {
        var sb = new StringBuilder();

        if (scopeLevel > 0)
        {
            sb.Append(' ', scopeLevel * 2);
        }

        sb.Append($"{GetShortLogLevelString(logLevel)} ");

        if (!string.IsNullOrEmpty(name))
        {
            sb.Append($"{name} ");
        }

        if (eventId.Id != 0)
        {
            sb.Append($"[{eventId.Id}]: ");
        }

        if (!string.IsNullOrEmpty(message))
        {
            sb.Append(message);
        }

        if (exception != null)
        {
            sb.Append($"\n{exception}");
        }

        return sb.ToString();
    }

    private static string GetShortLogLevelString(LogLevel level)
    {
        switch (level)
        {
            case LogLevel.Trace: return "Trace";
            case LogLevel.Debug: return "Debug";
            case LogLevel.Information: return "Info ";
            case LogLevel.Warning: return "Warn ";
            case LogLevel.Error: return "Error";
            case LogLevel.Critical: return "Crit ";
            case LogLevel.None: return "None ";
            default: throw new Exception("invalid");
        }
    }
}
```

File: Divergic.Logging.Xunit/ScopeWriter.cs [hunk 3]
```diff
@@ -52,7 +52,8 @@ public void Dispose()
 
         private static string BuildPadding(int depth)
         {
-            return new string(' ', depth * DefaultFormatter.PaddingSpaces);
+            var config = new LoggingConfig();
```
Comment (roryprimrose) [inline Divergic.Logging.Xunit/ScopeWriter.cs]: This produces a tight coupling between the ScopeWriter and the LoggingConfig class. The other issue here is that it ignores any custom config object that the consumer was wanting to use. This parameter needs to be passed into the constructor of the ScopeWriter.

File: Divergic.Logging.Xunit/LoggingConfig.cs (hunk 2 above)
Comment (dshe) [reply Divergic.Logging.Xunit/LoggingConfig.cs]: "Formatting is an action that in itself is not configuration." True. What about changing the class name of the class from LoggingConfig to LoggingProcessor/Outputter/Adapter/Generator/Producer? Then you could claim the class has a single, although more general, responsibility. Honoring SRP would sacrifice simplicity, which I believe is more valuable.

File: Divergic.Logging.Xunit/ScopeWriter.cs (hunk 3 above)
Comment (dshe) [reply Divergic.Logging.Xunit/ScopeWriter.cs]: BuildPadding(int depth) seems like it should be moved to the "LoggingProcessor" class.

File: Divergic.Logging.Xunit/LoggingConfig.cs (hunk 2 above)
Comment (roryprimrose) [reply Divergic.Logging.Xunit/LoggingConfig.cs]: Sorry, been busy. I take your point however I like the scope of the config types to describe config as it is the most appropriate description for that purpose in the API. The workaround I provided still gives you the simplicity in the calling implementation.

**Comment by dshe** [comment]
I have created my own logger. Please have a look:

https://github.com/dshe/XLogger

**Comment by fgheysels** [comment]
I had the same problem in a  testproject which targets netcoreapp3.1, using xunit 2.4.1 and Divergic.Logging.Xunit 3.4.0

I've switched to MXLogger (1.0.6), and my tests run without unexpected exceptions :)

**Pull request closed without merging**